Update containerd to default storage and add support for migration

dmcgowan opened the pull request:
Update storage driver selection to use containerd when no storage driver is provide.

Adds a new feature flag which will migrate to containerd snapshotter if there are no containers. To make layer migration efficient, only overlay2 and vfs is currently supported, less efficient options could be added later. The migration should be considered experimental and opt in.

Follow-ups to make it non-experimental
- Configurable threshold
- Configurable compression
- Better fallback/cleanup logic when migration fails

tianon commented on cmd/dockerd/config.go:
```diff
@@ -28,6 +28,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) error {
 	flags.StringVar(&conf.ExecRoot, "exec-root", conf.ExecRoot, "Root directory for execution state files")
 	flags.StringVar(&conf.ContainerdAddr, "containerd", "", "containerd grpc address")
 	flags.BoolVar(&conf.CriContainerd, "cri-containerd", false, "start containerd with cri")
+	flags.Var(opts.NewNamedSetOpts("feature", conf.Features), "feature", "Enable feature in the daemon")
```
I like what you're doing in this PR, but I have to admit this might be my favorite part :joy: :sob: :heart:

dmcgowan replied:
I could split this out

dmcgowan commented:
Moving this out of draft. There are a couple places we should add configurability and maybe some decisions that could be made differently, but this will be opt in to start and we can try it out. In testing using zstd made the migration 10x faster than gzip, we should consider using that by default for this.

tianon commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -2746,3 +2747,65 @@ func (s *DockerDaemonSuite) TestFailedPluginRemove(c *testing.T) {
+	).Assert(c, icmd.Success)
+	/*
+			cli.Docker(
+				cli.Args("run", "-d", "--name", "top1", "-p", "1234:80", "--restart", "always", "busybox:latest", "top"),
```
Is all this commented code intentionally present? :eyes:

dmcgowan replied:
I thought I ripped all the `integration-cli` stuff out but some was lurking in another commit 😜

dmcgowan committed:
Add containerd migration to daemon startup

Add layer migration on startup
Use image size threshold rather than image count
Add daemon integration test
Add test for migrating to containerd snapshotters
Add vfs migration
Add tar export for containerd migration
Add containerd migration test with save and load

Signed-off-by: Derek McGowan <derek@mcg.dev>

tianon commented:
> In testing using zstd made the migration 10x faster than gzip, we should consider using that by default for this.

I'm assuming you meant using zstd by default _just_ for the migration and not changing the default for pushing, but https://github.com/moby/moby/pull/48328 is related to the latter (linking so the discussions are connected).

dmcgowan committed:
Enable containerd snapshotters by default

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Fix containerd image count

Ensure image count returned by containerd image service only includes
the count of unique images.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Update migration test to use graphdriver env

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Check for snapshotter plugin availability

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Fix windows test graphdriver setting

Signed-off-by: Derek McGowan <derek@mcg.dev>

thaJeztah reviewed (commented):
looks like I started writing some comments (no full review) that I didn't submit yet

dmcgowan commented on daemon/containerd/service.go:
```diff
@@ -98,7 +98,15 @@ func (i *ImageService) CountImages(ctx context.Context) int {
 	}
 
-	return len(imgs)
+	uniqueImages := map[digest.Digest]struct{}{}
+	for _, i := range imgs {
+		dgst := i.Target().Digest
+		if _, ok := uniqueImages[dgst]; !ok {
+			uniqueImages[dgst] = struct{}{}
+		}
+	}
+
+	return len(uniqueImages)
```
"CountImages returns the number of images stored by ImageService" -  this is the same as `docker image ls` might return since that would be accounting for unique reference namespace. What we care about here is the actual number of images stored though, not number of references.

vvoland commented on daemon/daemon.go:
```diff
@@ -836,11 +861,24 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+		log.G(ctx).WithField("max_size", migrationThreshold).Info("(Experimental) Migration to containerd is enabled, driver will be switched to snapshotter after migration is complete")
+	}
+	if config.Features["containerd-snapshotter"] {
+		log.G(ctx).Warn(`"containerd-snapshotter" is now the default and no longer needed to be set`)
```
I think we should also handle a case where both containerd-migration is enabled and containerd snapshotter is explicitly disabled.

It's unclear what the expected behavior is in such case, so we should just error out

```suggestion
		log.G(ctx).Warn(`"containerd-snapshotter" is now the default and no longer needed to be set`)
	}
	if enabled, ok := config.Features["containerd-snapshotter"]; ok && !enabled{
		if config.Features["containerd-migration"] {
			// hard failure
		}
```

vvoland commented on daemon/daemon.go:
```diff
@@ -1157,14 +1190,121 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
-		// used above to run migration. They could be initialized in ImageService
-		// if migration is called from daemon/images. layerStore might move as well.
-		d.imageService = images.NewImageService(imgSvcConfig)
+		// If no containers are running, check whether can migrate image service
```
```suggestion
		// If no containers are present, check whether can migrate image service
```

This shouldn't happen if there are container created, but are not running right?

dmcgowan replied:
Updated, the status of the container should not make a difference for this check

vvoland commented on daemon/daemon.go:
```diff
@@ -836,11 +861,24 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+				return true
+			}
+			log.G(ctx).WithField("driver", driver).Warn("TEST_INTEGRATION_USE_GRAPHDRIVER is set but graphdriver is not registered")
+			return false
```
Shouldn't this be a hard failure? Usage of graphdrivers was explicitly requested (`TEST_INTEGRATION_USE_GRAPHDRIVER != ""`), but the config specifies an unknown driver.

dmcgowan replied:
Updated

vvoland commented on daemon/daemon.go:
```diff
@@ -1157,14 +1190,121 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+
+			}
+
+			if totalSize <= migrationThreshold {
```
https://github.com/moby/moby/blob/3023608e18e86df5d0e5a48def4fc0fe86992df4/daemon/daemon.go#L877

Looks like this (`totalSize <= math.MaxInt64`) would always be true regardless if there are images left or not?

vvoland replied:
Can we simplify this whole `migrationThreshold` to just a boolean `doMigrate := true/false` to simplify this for now, especially since we don't support setting the threshold?

dmcgowan replied:
This value can also be set to 0 though, indicating that we should only switch to containerd when image count goes to zero.

vvoland replied:
Right but it's only for windows in one specific case:

https://github.com/moby/moby/blob/3023608e18e86df5d0e5a48def4fc0fe86992df4/daemon/daemon.go#L1088-L1103

I think this should default to `0` so that we only migrate once users reach empty state.

https://github.com/dmcgowan/docker/blob/8f7bd1f547b13ef5fbfa5b5d9f5a3f3b287bf67e/daemon/daemon.go#L878

dmcgowan replied:
The migration will need more thought and testing before no longer being experimental. Changing the default threshold to 0 would change the logic though and would necessitate adding the threshold in anyway in order to test it. Can add the threshold in a follow up, it is trivial, but it is experimental and not the most important aspect of this PR.

dmcgowan committed:
Update graphdriver check logic to account for disabling of snapshotter

Ensure migration is never attempted multiple times.

Signed-off-by: Derek McGowan <derek@mcg.dev>

vvoland commented:
A minor issue is that because of https://github.com/moby/moby/issues/49784, repulling the migrated images doesn't produce a "complete" image:


<img width="998" height="202" alt="image" src="https://github.com/user-attachments/assets/a06ede4d-5050-4905-a550-a935b41dbf35" />

dmcgowan committed:
Add environment variable to define the threshold

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Fix symlink evaluation to a directory that may not exist

During the arm64 tests, the rootfs directory does not seem to exist when
this test is run and will cause a failure when using snapshotter.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Use native snapshotter for integration tests and run

Signed-off-by: Derek McGowan <derek@mcg.dev>

thaJeztah commented on daemon/daemon.go:
```diff
@@ -241,29 +243,39 @@ func (daemon *Daemon) restore(cfg *configStore) error {
+	return driverContainers, nil
+}
+
+func (daemon *Daemon) restore(cfg *configStore, containers map[string]*container.Container) error {
```
Same thing wondering here (about context). If needed, we could still make it a `context.WithoutCancel` at the start of the function, or pass a `context.TODO()`.

thaJeztah commented on daemon/daemon.go:
```diff
@@ -33,6 +33,7 @@ import (
 	"github.com/containerd/log"
 	"github.com/distribution/reference"
 	dist "github.com/docker/distribution"
+	"github.com/dustin/go-humanize"
```
Curious if this package is consistent with `github.com/docker/go-units` which we use elsewhere. (I know it has issues, but I think we use that package in most places).

thaJeztah commented on go.mod:
```diff
@@ -162,7 +162,7 @@ require (
 	github.com/dimchansky/utfbom v1.1.1 // indirect
 	github.com/docker/libtrust v0.0.0-20150526203908-9cbd2a1374f4 // indirect
-	github.com/dustin/go-humanize v1.0.0 // indirect
+	github.com/dustin/go-humanize v1.0.0
```
Related to the above; evidently we did not use this package before, so perhaps we should look if `docker/go-units` should be preferred for this.

thaJeztah reviewed (commented):
made a pass, but will probably need to go through this again

dmcgowan committed:
Fix hardlink handling in containerd snapshot remap

When files are hardlinked, the inodes only need to be chowned once.

Signed-off-by: Derek McGowan <derek@mcg.dev>

dmcgowan committed:
Add context to restore and load containers

Signed-off-by: Derek McGowan <derek@mcg.dev>

austinvazquez reviewed (approved):
LGTM; just a few nits and q's from me. Not worth blocking unless you happen to find yourself under the hood again.

thaJeztah commented on daemon/daemon.go:
```diff
@@ -1033,50 +1087,49 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
 	// Unix platforms however run a single graphdriver for all containers, and it can
 	// be set through an environment variable, a daemon start parameter, or chosen through
 	// initialization of the layerstore through driver priority order for example.
-	driverName := os.Getenv("DOCKER_DRIVER")
```
This looks like a breaking change; the `DOCKER_DRIVER` env-var has been in the daemon for a long time (https://github.com/moby/moby/commit/aea6001baf0cd0fb20a607e37e4379bf644b28fd); while it was originally for testing, it is used externally. The `DOCKER_GRAPHDRIVER` one was only used in test-code, so we should probably remove it from the go-code (and map it in our scripts or makefiles if needed); 

```bash
git grep -E 'DOCKER_GRAPHDRIVER|DOCKER_DRIVER' -- '*.go'
daemon/daemon.go:               driverName := os.Getenv("DOCKER_DRIVER")
daemon/daemon.go:                       logrus.Infof("Setting the storage driver from the $DOCKER_DRIVER environment variable (%s)", driverName)
testutil/daemon/daemon.go:      storageDriver := os.Getenv("DOCKER_GRAPHDRIVER")
```

We should continue to support the env-var, because it's used in quite some places;

- https://github.com/gitlabhq/gitlabhq/blob/9d33bc1382513f897a1ad9400591b06446b1fcc5/doc/ci/testing/code_quality_codeclimate_scanning.md?plain=1#L537-L555
- https://github.com/coherence-community/oracle-bedrock/blob/b95419ac5d86b99086369a5d45041c90d2f73e38/bedrock-runtime-docker/src/main/java/com/oracle/bedrock/runtime/docker/Docker.java#L599-L613
